stdlib/src/builtin/string.mojo

```diff
@@ -2100,3 +2174,169 @@ fn _calc_format_buffer_size[type: DType]() -> Int:
+                        )
+                        entries.append(curren_entry)
+                        start = None
+                        continue
+                    raise (
+                        "there is a single curly { left unclosed or unescaped"
+                    )
+                else:
+                    start = i
+                continue
+            if format_src[i] == "}":
+                if start:
+                    var start_value = start.value()
+                    var current_entry = Self(
+                        first_curly=start_value, last_curly=i, field=None
+                    )
+                    if i - start_value != 1:
+                        var field = format_src[start_value + 1 : i]
+                        try:
+                            # field is a number for manual indexing:
+                            var number = int(field)
+                            current_entry.field = number
+                            if number >= len_pos_args or number < 0:
+                                raised_manual_index = number
+                                break
+                            manual_indexing_count += 1
+                        except e:
+                            debug_assert(
+                                "not convertible to integer" in str(e),
+                                "Not the expected error from atol",
+                            )
+                            # field is an keyword for **kwargs:
+                            current_entry.field = field
+                            kwargs_indexing_count += 1
+                            raised_kwarg_field = field
+                            break
+                    else:
+                        # automatic indexing
+                        # current_entry.field is already None
+                        if automatic_indexing_count >= len_pos_args:
+                            raised_automatic_index = automatic_indexing_count
+                            break
+                        automatic_indexing_count += 1
+                    entries.append(current_entry^)
+                    start = None
+                else:
+                    # python escapes double curlies
+                    if (i + 1) < len(format_src):
+                        if format_src[i + 1] == "}":
+                            var curren_entry = Self(
+                                first_curly=i, last_curly=i + 1, field=True
+                            )
+                            entries.append(curren_entry)
+                            skip_next = True
+                            continue
+                    # if it is not an escaped one, it is an error
+                    raise (
+                        "there is a single curly } left unclosed or unescaped"
+                    )
+
+        if raised_automatic_index:
+            raise "Automatic indexing require more args in *args"
```

Comment:
non-blocking: I'm not sure I understand why you moved this here. The architecture as I see it should be is the following: `_FormatCurlyEntry` is a helper class that takes _only_ the template string, parses it, and returns a data structure that `String.format` can easily use to perform its own checks and the actual substitution. So `_FormatCurlyEntry.create_entries` raises only if there is something fundamentally wrong with the template string (e.g. both manual and automatic indexing is used, there are inbalanced braces, etc). The other errors are the responsibility of `String.format`.

This is important, because it would be good to retain this separation of concern as we build out other features, for example `"{foo:0.5}".format(435.563244535)`. WDYT?

Reply:
Yes, that makes sense, sorry moved too much there!
If we return only a `List[_FormatCurlyEntry]`, we would need to iterate it two times in `format`.
We'll probably need to return the counts with the list, in order to raise errors before starting.



`format` need to raise for `count_*>=len(args)`  and not `create_entries`?

:+1: All for re-use and separations

Reply:
@laszlokindrat should move the `raises` in this patch ?

Reply:
How about in a followup? I'm eager to land this, and a followup with this changes should be straightforward to review.